app/models/targeting.rb

```diff
@@ -0,0 +1,29 @@
+class Targeting < ActiveRecord::Base
+
+  belongs_to :user
+  belongs_to :bookmark
+
+  has_many :targeting_hosts, :dependent => :destroy
+  has_many :hosts, :through => :targeting_hosts
+  has_many :template_invocations, :dependent => :destroy
+
+  validate :bookmark_or_query
+
+  def resolve_hosts!
+    fail _("Cannot resolve hosts without a user") if user.nil?
+    fail _("Cannot resolve hosts without a bookmark or search query") if bookmark.nil? && search_query.blank?
+    query = bookmark.try(:query) || search_query
```

Comment:
so i guess one question that is not answered is 'when does invocation happen?'

If i create a targeting and point it at a bookmark, would i want that query saved right then or in the future (assuming its scheduled to run in the future).  I would think by default I would want it resolved in the future with whatever that bookmark is set to at that time.  So that would just mean saving the query when the host resolution happens.